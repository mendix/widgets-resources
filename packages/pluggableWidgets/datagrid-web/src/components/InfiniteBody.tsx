import { createElement, CSSProperties, ReactElement, ReactNode, useCallback, useState } from "react";
import classNames from "classnames";

interface InfiniteBodyProps {
    children: ReactNode;
    hasMoreItems: boolean;
    setPage?: (computePage: (prevPage: number) => number) => void;
    isInfinite: boolean;
    style?: CSSProperties;
}

export function InfiniteBody({ children, hasMoreItems, setPage, isInfinite, style }: InfiniteBodyProps): ReactElement {
    const [bodySize, setBodySize] = useState(0);

    const trackScrolling = useCallback(
        e => {
            /**
             * In Windows OS the result of first expression returns a non integer and result in never loading more, require floor to solve.
             */
            const bottom = Math.floor(e.target.scrollHeight - e.target.scrollTop) === Math.floor(e.target.clientHeight);
            if (bottom) {
                if (hasMoreItems && setPage) {
                    setPage(prev => prev + 1);
                }
            }
        },
        [hasMoreItems, setPage]
    );

    const calculateBodyHeight = useCallback(
        (ref: HTMLTableSectionElement): void => {
            if (ref && isInfinite && bodySize <= 0 && hasMoreItems) {
                setBodySize(ref.clientHeight - 30);
            }
        },
        [bodySize, isInfinite, hasMoreItems]
    );

    return (
        <div
            ref={calculateBodyHeight}
            className={classNames("table-content", isInfinite ? "infinite-loading" : "")}
            onScroll={isInfinite ? trackScrolling : undefined}
            role="rowgroup"
            style={isInfinite && bodySize > 0 ? { ...style, maxHeight: bodySize } : style}
        >
            {children}
        </div>
    );
}
