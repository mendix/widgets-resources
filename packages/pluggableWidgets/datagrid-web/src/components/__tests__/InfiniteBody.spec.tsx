import { shallow } from "enzyme";
import { createElement } from "react";
import { InfiniteBody } from "../InfiniteBody";

describe("Infinite Body", () => {
    it("renders the structure correctly", () => {
        const component = shallow(createElement(InfiniteBody));

        expect(component).toMatchSnapshot();
    });
});
