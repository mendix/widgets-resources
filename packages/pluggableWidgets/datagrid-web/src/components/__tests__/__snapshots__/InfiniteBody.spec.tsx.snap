// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Infinite Body renders the structure correctly 1`] = `
<div
  className="table-content"
  role="rowgroup"
/>
`;
